inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/AbstractKotlinElement.kt

```diff
@@ -94,7 +113,18 @@ internal abstract class AbstractKotlinElement<T : KotlinNativeElement>(
     }
 
     override fun isFinal() = if (annotatedInfo is KSDeclaration) {
-        !annotatedInfo.isOpen() || annotatedInfo.modifiers.contains(Modifier.FINAL)
+        if (annotatedInfo.modifiers.contains(Modifier.FINAL)) {
+            true
+        } else if (annotatedInfo.isOpen()) {
+            false
+        } else {
+            // ksp does not see when all-open opens up these classes
+            // https://github.com/micronaut-projects/micronaut-core/issues/9426
+            // this logic is similar to what all-opens does.
+            val cl = annotatedInfo as? KSClassDeclaration ?: annotatedInfo.parentDeclaration as? KSClassDeclaration
+            cl == null || !visitorContext.elementFactory.newClassElement(cl)
```

Comment:
Why do you need to create a new element? You should be able to access the annotation from the superclass `AbstractAnnotationElement`

Reply:
it's not always the right annotations. e.g. when the ClassElement is loaded from a field, it will contain the annotations of the field too.

Reply:
I don’t understand what are you talking about. If the class is annotated with one of those annotations it should not indicate final class. I don’t see anything that prevents to use ordinary class’ annotation metadata

Reply:
for example `test final modifier on factory with AOP advice doesn't compile` fails if i use `this` instead of a new class element, because the method is annotated with `@Mutating` so the return type ClassElement inherits that annotation from the method even though the class itself does not have the annotation.

Reply:
to be clear in that test, with `this`, the processor incorrectly thinks the all-open plugin would open up `MyBean.`

Reply:
I will check

Reply:
Corrected